### PR TITLE
Hardhat deploy plugin and deployment of the layer 1 bits.

### PR DESCRIPTION
**Pull request by StefanIliev545**
### Why is this change needed?

In order to make hardhat capable of deploying to our environments.

This one does not deploy to L2 as the wallet extension bits are a prerequisite. 

### What changes were made as part of this PR:

Hardhat deploy plugin is added. 
The contracts for L1 and testnet specific ones are added.
Dockerfile for the hardhat deployer and a script replacing the one using the old contract deployer.

### Definition of done

- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated 
- [ ] End-to-end tests run if required (see below)

### End-to-end tests

Running the end-to-end tests in the [obscuro-test repo](https://github.com/obscuronet/obscuro-test) is currently at the 
discretion of the developer prior to merging a PR. The intention is not to slow down development by having the longer 
running end-to-end tests as a PR gate run on each branch commit etc. To manually trigger a run pre-merge;

- Go to [run_local_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_local_tests.yml)
- Click the "Run workflow" button
- Use the main branch of obscuro-test 
- Enter the name of the PR branch for go-obscuro
- Run the workflow

The run takes approximately 20 minutes and any failure will be reported in the workflow output. Should the tests fail 
the docker container output for the local testnet and all test artifacts will be added to the run and can be downloaded
for debugging and analysis, with a retention period of 2 days. 

Note that every PR merge to main will also trigger a run of post-merge tests, so even if you do not manually trigger 
pre-merge, tests will be run post-merge. The intention being to catch any issues fast on main to allow for quick 
resolution. The post-merge test output can be seen at 
[run_merge_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_merge_tests.yml) and will show both 
the PR number and author in the list of runs.

### CODE REVIEW

File: contracts/deployment_scripts/testnet/001_deploy_tokens.ts
Comment (otherview) [inline contracts/deployment_scripts/testnet/001_deploy_tokens.ts:30]: Similarly, this contract can be executed with:
```
await lock.unlockTime()
```

File: contracts/hardhat.config.ts
Comment (otherview) [inline contracts/hardhat.config.ts:35]: Are these requirements from the deploy plugin ?

File: contracts/hardhat.config.ts
Comment (otherview) [inline contracts/hardhat.config.ts:57]: Is reading from env and from file a requirement ?
Could the networks.json file be described in this file ?

File: testnet/testnet-deploy-l1-contracts.sh
Comment (otherview) [inline testnet/testnet-deploy-l1-contracts.sh:80]: Can a comment be added here ?

File: contracts/deployment_scripts/layer1/001_deploy_management_contract.ts
Comment (otherview) [inline contracts/deployment_scripts/layer1/001_deploy_management_contract.ts:2]: Not seeing the value of using the deploy package. Can you ping across some examples ? 
I would prefer to avoid future-proofing, better to address the needs as we have them.

**Review by otherview — COMMENTED** [review]
Overall looks good.
The github workflows I believe need to be updated.
I think it's possible to reduce some components.

File: testnet/hardhatdeployer.Dockerfile
Comment (StefanIliev545) [reply testnet/hardhatdeployer.Dockerfile:37]: The volume is so that the result of the deployment is persisted. This is to be used later by the L2 deployment that needs to link the bridges. The plugin saves a ton of stuff in this volume. Also if you run twice it will skip deploying scripts that have already been deployed (but will verify everything is there and so on)

File: contracts/deployment_scripts/testnet/001_deploy_tokens.ts
Comment (StefanIliev545) [reply contracts/deployment_scripts/testnet/001_deploy_tokens.ts:12]: Mainly that the plugin persists what it has deployed, makes it accessible by name in other deployments and knows on which network it is deployed so it can manage complicated bits. Also it has some fail safes to prevent deploying when stuff is out of order.

File: contracts/hardhat.config.ts
Comment (StefanIliev545) [reply contracts/hardhat.config.ts:35]: Not required, but it's nice. Essentially allows you to declare which key belongs to which named account. Otherwise the plugin introduces it.

File: contracts/hardhat.config.ts
Comment (StefanIliev545) [reply contracts/hardhat.config.ts:57]: I've had the networks json described in here, but it felt like it was bloating the file. When default configs for testnet, simulation and so on are added at least.

File: contracts/deployment_scripts/layer1/001_deploy_management_contract.ts
Comment (StefanIliev545) [reply contracts/deployment_scripts/layer1/001_deploy_management_contract.ts:2]: At the stage of only deploying to L1 there isn't much of a value, apart from the checks if contracts exist and not deploying multiple times, but when deploying to L1 and L2 and requiring to maintain connection and deployed state across two networks it makes it easier due to being able to define companion networks and then calling functions like

```
layer1.execute(...
layer2.execute(...
```

The HRE is not really good by itself for multiple networks.

File: contracts/deployment_scripts/testnet/001_deploy_tokens.ts
Comment (StefanIliev545) [reply contracts/deployment_scripts/testnet/001_deploy_tokens.ts:12]: In the example you gave, hardhat when started instantiates some singleton stuff to the network it was started with and caches a ton of stuff so `hre.ethers` always points to a single network. We'd have to manually instantiate stuff when we want to orchestrate cross network deployments. But yeah, we can do it entirely through ethers, it would just be additional code and manually managing what account is connected and so on.

File: testnet/hardhatdeployer.Dockerfile
Comment (otherview) [reply testnet/hardhatdeployer.Dockerfile:37]: In the past, I've used volume to quickly test changes without rebuilding the whole container, by mounting the volume on the host. I think if that's the approach that's fine, even desirable.

While there is a great caching benefit of using volume it's not desirable to have stateful containers.
Each execution should not read any state from previous executions.

**Comment by otherview** [comment]
Just capturing the offline discussion here
- Move the TS to JS
- Remove the Volume
- Keeping the HH Deploy plugin
- Execution is now able to receive it it's deploying to L1, L2 or both

Please update if something is missing @StefanIliev545  :)

File: contracts/hardhat.config.ts
Comment (otherview) [reply contracts/hardhat.config.ts:57]: Cool, got it. What about the reading from env and file ? is the file read connected with the docker volume ?

**Review by otherview — COMMENTED** [review]
Looks good.

Before merging please make sure it works with deploying to dev-testnet and please check with @moraygrieve if it does not impact the obscuro-tests. It shouldn't, but worth checking

**Review by tudor-malene — APPROVED** [review]
lgtm (if it works)

I haven't looked in detail, but I trust Pedro has.

**Pull request merged**
